python/ray/train/tests/test_trainer_restore.py

```diff
@@ -161,30 +174,34 @@ def test_gbdt_trainer_restore(ray_start_6_cpus, tmpdir, trainer_cls):
             num_workers=2, trainer_resources={"CPU": 0}, resources_per_worker={"CPU": 1}
         ),
         run_config=RunConfig(
-            local_dir=str(tmpdir),
             name=exp_name,
             checkpoint_config=CheckpointConfig(
                 num_to_keep=1, checkpoint_frequency=1, checkpoint_at_end=False
             ),
-            callbacks=[FailureInjectionCallback(num_iters=2)],
+            callbacks=[FailureInjectionCallback(fail_marker_path, num_iters=2)],
         ),
         num_boost_round=5,
     )
     with pytest.raises(TrainingFailedError):
         result = trainer.fit()
 
-    trainer = trainer_cls.restore(str(tmpdir / exp_name), datasets=datasets)
+    trainer = trainer_cls.restore(str(tmp_path / exp_name), datasets=datasets)
     result = trainer.fit()
     assert not result.error
     assert result.metrics["training_iteration"] == 5
     assert result.metrics["iterations_since_restore"] == 3
-    assert tmpdir / exp_name in result.log_dir.parents
+    assert tmp_path / exp_name in Path(result.path).parents
 
 
 @pytest.mark.parametrize("trainer_cls", [TransformersTrainer])
-def test_trainer_with_init_fn_restore(ray_start_4_cpus, tmpdir, trainer_cls):
+def test_trainer_with_init_fn_restore(
+    ray_start_4_cpus, tmp_path, trainer_cls, monkeypatch
+):
     """Tests restore for data parallel trainers that take in a `train_init` function
     and config. Success criteria: same as for data parallel trainers."""
+    # TODO(justinvyu): Failure injection callback doesn't work very well to test this.
+    pytest.skip("Re-enable after coming up with a better way to inject a failure.")
+    monkeypatch.setenv("RAY_AIR_LOCAL_CACHE_DIR", str(tmp_path))
```

Comment:
This test is failing for some complicated reasons:
1. We inject a failure in a callback. Errors in callbacks get handled in different ways.
2. If we fail during `on_trial_save`, the error gets raised immediately WITHOUT calling `tune_controller.checkpoint` one last time. This means that the checkpoint gets processed ->  checkpoint 00001 gets deleted, but the experiment state still thinks checkpoint 00001 is the latest one. This is a consequence of raising exactly during `on_trial_save` + `num_to_keep=1`.
3. I tried failing on the next iteration's `on_trial_result`, but that also didn't work. The trial advanced to the next iteration for some reason.

So, I'm skipping for these (legacy) trainers for now.

Reply:
Fine with me to skip the test, but for the case where `on_trial_save` fails (in a custom callback or so), should we make sure this works? (Can be in a follow-up)

Reply:
Sounds good. I'm not sure why this worked before actually, would be good to get to the bottom of this in a follow-up.

Tracking that here https://github.com/ray-project/ray/issues/38734